modules/currencies/src/lib.rs

```diff
@@ -637,12 +657,332 @@ impl<T: Config> MultiReservableCurrency<T::AccountId> for Pallet<T> {
+			currency_id: asset_id,
+			from: source.clone(),
+			to: dest.clone(),
+			amount: transferred_amount,
+		});
+		Ok(transferred_amount)
+	}
+}
+
+impl<T: Config> fungibles::Unbalanced<T::AccountId> for Pallet<T> {
+	fn set_balance(asset_id: Self::AssetId, who: &T::AccountId, amount: Self::Balance) -> DispatchResult {
+		match asset_id {
+			CurrencyId::Erc20(_) => Err(Error::<T>::Erc20InvalidOperation.into()),
+			id if id == T::GetNativeCurrencyId::get() => {
+				<T::NativeCurrency as fungible::Unbalanced<_>>::set_balance(who, amount)
+			}
+			_ => <T::MultiCurrency as fungibles::Unbalanced<_>>::set_balance(asset_id, who, amount),
+		}
+	}
+
+	fn set_total_issuance(asset_id: Self::AssetId, amount: Self::Balance) {
+		match asset_id {
+			CurrencyId::Erc20(_) => {}
+			id if id == T::GetNativeCurrencyId::get() => {
+				<T::NativeCurrency as fungible::Unbalanced<_>>::set_total_issuance(amount)
+			}
+			_ => <T::MultiCurrency as fungibles::Unbalanced<_>>::set_total_issuance(asset_id, amount),
+		}
+	}
+}
+
+impl<T: Config> fungibles::InspectHold<T::AccountId> for Pallet<T> {
+	fn balance_on_hold(asset_id: Self::AssetId, who: &T::AccountId) -> Self::Balance {
+		match asset_id {
+			CurrencyId::Erc20(_) => <Self as MultiReservableCurrency<_>>::reserved_balance(asset_id, who),
+			id if id == T::GetNativeCurrencyId::get() => {
+				<T::NativeCurrency as fungible::InspectHold<_>>::balance_on_hold(who)
+			}
+			_ => <T::MultiCurrency as fungibles::InspectHold<_>>::balance_on_hold(asset_id, who),
+		}
+	}
+
+	fn can_hold(asset_id: Self::AssetId, who: &T::AccountId, amount: Self::Balance) -> bool {
+		match asset_id {
+			CurrencyId::Erc20(_) => {
+				amount.is_zero()
+					|| <Self as fungibles::InspectHold<_>>::balance_on_hold(asset_id, who)
+						.checked_add(&amount)
+						.is_some()
+			}
+			id if id == T::GetNativeCurrencyId::get() => {
+				<T::NativeCurrency as fungible::InspectHold<_>>::can_hold(who, amount)
+			}
+			_ => <T::MultiCurrency as fungibles::InspectHold<_>>::can_hold(asset_id, who, amount),
+		}
+	}
+}
+
+impl<T: Config> fungibles::MutateHold<T::AccountId> for Pallet<T> {
+	fn hold(asset_id: Self::AssetId, who: &T::AccountId, amount: Self::Balance) -> DispatchResult {
+		match asset_id {
+			CurrencyId::Erc20(_) => <Self as MultiReservableCurrency<_>>::reserve(asset_id, who, amount),
+			id if id == T::GetNativeCurrencyId::get() => {
+				<T::NativeCurrency as fungible::MutateHold<_>>::hold(who, amount)
+			}
+			_ => <T::MultiCurrency as fungibles::MutateHold<_>>::hold(asset_id, who, amount),
+		}
+	}
+
+	fn release(
+		asset_id: Self::AssetId,
+		who: &T::AccountId,
+		amount: Self::Balance,
+		best_effort: bool,
+	) -> Result<Self::Balance, DispatchError> {
+		match asset_id {
+			CurrencyId::Erc20(_) => {
+				if amount.is_zero() {
+					return Ok(amount);
+				}
+				ensure!(
+					amount <= <Self as MultiReservableCurrency<_>>::reserved_balance(asset_id, who) || best_effort,
```

Comment:
```suggestion
					best_effort || amount <= <Self as MultiReservableCurrency<_>>::reserved_balance(asset_id, who),
```

do `best_effort`check first, to reduce one database read if it's true